// commandpost/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util = require("util");
var ErrorReason;
(function (ErrorReason) {
    ErrorReason["UnsupportedFormatArgument"] = "unsupported format";
    ErrorReason["ArgumentsRequired"] = "1 or more arguments required";
    ErrorReason["ArgumentRequired"] = "argument required";
    ErrorReason["ParameterCantPlacedAfterVariadic"] = "parameter can not placed after variadic parameter";
    ErrorReason["ParameterCannPlacedAfterOptional"] = "required parameter is not placed after optional parameter";
    ErrorReason["UnknownOption"] = "unknown option";
    ErrorReason["OptionNameMismatch"] = "short or long option name mismatch";
    ErrorReason["OptionValueRequired"] = "option value required";
})(ErrorReason = exports.ErrorReason || (exports.ErrorReason = {}));
var CommandpostError = /** @class */ (function () {
    function CommandpostError(params) {
        this.params = params;
        Error.captureStackTrace(this, this.constructor);
    }
    Object.defineProperty(CommandpostError.prototype, "name", {
        get: function () {
            return this.constructor.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CommandpostError.prototype, "message", {
        get: function () {
            return this.params.message;
        },
        enumerable: true,
        configurable: true
    });
    return CommandpostError;
}());
exports.CommandpostError = CommandpostError;
util.inherits(CommandpostError, Error);
