// commandpost/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
try {
    // optional
    require("source-map-support").install();
}
catch (e) {
}
try {
    if (typeof Promise === "undefined") {
        require("es6-promise").polyfill();
    }
}
catch (e) {
}
var command_1 = require("./command");
exports.Command = command_1.default;
var option_1 = require("./option");
exports.Option = option_1.default;
var argument_1 = require("./argument");
exports.Argument = argument_1.default;
var error_1 = require("./error");
exports.CommandpostError = error_1.CommandpostError;
exports.ErrorReason = error_1.ErrorReason;
/**
 * Create new top level command.
 * @param cmdName
 * @returns {Command<Opt, Arg>}
 */
function create(cmdName) {
    return new command_1.default(cmdName);
}
exports.create = create;
/**
 * exec parsing and call callback.
 * @param cmd it created by create function.
 * @param argv pass process.argv
 * @returns {Promise<{}>}
 */
function exec(cmd, argv) {
    return Promise
        .resolve(null)
        .then(function () {
        argv = argv.slice(2);
        // cmd.parse throw an exception often.
        return cmd.parse(argv);
    });
}
exports.exec = exec;
